File: src/infinicore/ops/spmm/spmm.cc
#include "infinicore/ops/spmm.hpp"

#include "../../utils.hpp"

namespace infinicore::op {
INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(SpMM);

SpMM::SpMM(Tensor c, const SpMat &a, const Tensor &b, float alpha, float beta) {
    INFINICORE_ASSERT(a);
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(c, a->values(), a->crow_indices(), a->col_indices(), b);
    INFINICORE_GRAPH_OP_DISPATCH(c->device().getType(), c, a, b, alpha, beta);
}

void SpMM::execute(Tensor c, const SpMat &a, const Tensor &b, float alpha, float beta) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(SpMM, c, a, b, alpha, beta);
}

Tensor spmm(const SpMat &a, const Tensor &b, float alpha, float beta) {
    INFINICORE_ASSERT(a);
    auto c = Tensor::empty({a->rows(), b->size(1)}, a->dtype(), b->device());
    spmm_(c, a, b, alpha, beta);
    return c;
}

void spmm_(Tensor c, const SpMat &a, const Tensor &b, float alpha, float beta) {
    SpMM::execute(c, a, b, alpha, beta);
}

} // namespace infinicore::op

File: python/infinicore/ops/spmm.py
from infinicore.lib import _infinicore
from infinicore.spmat import SpMat
from infinicore.tensor import Tensor


def spmm(a, b, *, alpha=1.0, beta=0.0, out=None):
    if not isinstance(a, SpMat):
        raise TypeError("spmm expects a CSR SpMat as the left-hand side")

    if out is None:
        return Tensor(_infinicore.spmm(a._underlying, b._underlying, alpha, beta))

    _infinicore.spmm_(
        out._underlying,
        a._underlying,
        b._underlying,
        alpha,
        beta,
    )
    return out
